source/MRMesh/MRColorMapAggregator.cpp

```diff
@@ -0,0 +1,161 @@
+#include "MRColorMapAggregator.h"
+#include "MRGTest.h"
+
+namespace MR
+{
+
+template<typename Tag>
+void ColorMapAggregator<Tag>::setDefaultColor( const Color& color )
+{
+    defaultColor_ = color;
+}
+
+template<typename Tag>
+void ColorMapAggregator<Tag>::pushBack( const ColorMap& colorMap, const ElementBitSet& elementsBitSet )
+{
+    checkInputData_( colorMap, elementsBitSet );
+    dataSet_.push_back( { colorMap, elementsBitSet } );
+}
+
+template<typename Tag>
+void ColorMapAggregator<Tag>::insert( int i, const ColorMap& colorMap, const ElementBitSet& elementsBitSet )
+{
+    assert( i < dataSet_.size() );
+    checkInputData_( colorMap, elementsBitSet );
+    dataSet_.push_back( { colorMap, elementsBitSet } );
+}
+
+template<typename Tag>
+void ColorMapAggregator<Tag>::replace( int i, const ColorMap& colorMap, const ElementBitSet& elementsBitSet )
+{
+    assert( i >= 0 && i < dataSet_.size() );
+    checkInputData_( colorMap, elementsBitSet );
+    dataSet_[i] = { colorMap, elementsBitSet };
+}
+
+template<typename Tag>
+void ColorMapAggregator<Tag>::erase( int i, int n /*= 1*/ )
+{
+    assert( i >= 0 && i + n <= dataSet_.size() );
+    dataSet_.erase( dataSet_.begin() + i, dataSet_.begin() + n );
+
+    needUpdate_ = true;
+    if ( dataSet_.empty() )
+        colorMapSize_ = 0;
+}
+
+template<typename Tag>
+void ColorMapAggregator<Tag>::setMode( AggregateMode mode )
+{
+    if ( mode == mode_ )
+        return;
+    mode_ = mode;
+    needUpdate_ = true;
+}
+
+template<typename Tag>
+typename ColorMapAggregator<Tag>::ColorMap ColorMapAggregator<Tag>::aggregate()
+{
+    if ( needUpdate_ )
+        updateAggregated();
+    return aggregatedColorMap_;
+}
+
+template<typename Tag>
+void ColorMapAggregator<Tag>::checkInputData_( const ColorMap& colorMap, const ElementBitSet& elementsBitSet )
+{
+    assert( !colorMap.empty() );
+    assert( colorMap.size() == elementsBitSet.size() );
```

Comment:
better assert that `colorMap.size() >= elementsBitSet.size()`

Reply:
or even better `colorMap.size() >= elementsBitSet.last_valid()`, if find last valid is not very hard

Reply:
fixed